sidecar/src/service/runtime_info.rs

```diff
@@ -0,0 +1,139 @@
+    pub(crate) instance_id: InstanceId,
+}
+
+// TODO-EK: Can this be refactored more before merging? We may be able to encapsulate some of the
+// functionality in session_info to RuntimeInfo.
+impl RuntimeInfo {
+    /// Retrieves the `AppInstance` for a given service name and environment name.
+    ///
+    /// # Arguments
+    ///
+    /// * `service_name` - A string slice that holds the name of the service.
+    /// * `env_name` - A string slice that holds the name of the environment.
+    ///
+    /// # Returns
+    ///
+    /// * `SharedAppManualFut` - A struct that contains the shared future of the `AppInstance` and
+    ///   its completer.
+    pub(crate) fn get_app(&self, service_name: &str, env_name: &str) -> SharedAppManualFut {
+        let mut apps = self.lock_apps();
+        let key = (service_name.to_owned(), env_name.to_owned());
+        if let Some(found) = apps.get(&key) {
+            SharedAppManualFut {
+                app_future: found.clone(),
+                completer: None,
+            }
+        } else {
+            let (future, completer) = ManualFuture::new();
+            let shared = future.shared();
+            apps.insert(key, shared.clone());
+            SharedAppManualFut {
+                app_future: shared,
+                completer: Some(completer),
+            }
+        }
+    }
+    /// Shuts down the runtime.
+    /// This involves shutting down all the instances in the runtime.
+    pub(crate) async fn shutdown(self) {
+        #[cfg(feature = "tracing")]
+        info!(
+            "Shutting down runtime_id {} for session {}",
+            self.instance_id.runtime_id, self.instance_id.session_id
+        );
+
+        let instance_futures: Vec<_> = self
+            .lock_apps()
+            .drain()
+            .map(|(_, instance)| instance)
+            .collect();
+        let instances: Vec<_> = join_all(instance_futures).await;
+        let instances_shutting_down: Vec<_> = instances
+            .into_iter()
+            .map(|instance| {
+                tokio::spawn(async move {
+                    if let Some(instance) = instance {
+                        instance
+                            .telemetry
+                            .send_msg(TelemetryActions::Lifecycle(LifecycleAction::Stop))
+                            .await
+                            .ok();
+                        instance.telemetry_worker_shutdown.await;
+                    }
+                })
+            })
+            .collect();
+        future::join_all(instances_shutting_down).await;
+
+        #[cfg(feature = "tracing")]
+        debug!(
+            "Successfully shut down runtime_id {} for session {}",
+            self.instance_id.runtime_id, self.instance_id.session_id
+        );
+    }
+
+    // TODO-EK: Reduce the type complexity before merging
+    #[allow(clippy::type_complexity)]
+    // TODO-EK - Investigate why this needs to be exposed this way before merging
```

Comment:
The goal is simply simplifying the code a bit whether every access is multiple lines, see https://github.com/DataDog/libdatadog/pull/392#discussion_r1565757801 for context.

Reply:
Reducing the type complexity here sort of defeats the point of just being a simple helper function.

Reply:
@bwoebi your comments are from an outdated version of this file, so I'm not sure if still applies. 

The complexity isn't due to the `MutexGuard`, it's due to the app HashMap type of `HashMap<(String, String), Shared<ManualFuture<Option<AppInstance>>>>`. I created the type `AppMap` and use that throughout `RuntimeInfo` and was able to remove the clippy ignore. 

I think we should strive to ignore linter errors only when it is absolutely necessary. I can't speak for @bantonsson on https://github.com/DataDog/libdatadog/pull/392#discussion_r1565757801, but I would argue that a helper function isn't all that helpful if you have to suppress valid linter errors. The current implementation resolves the clippy error and maintains the helpfulness Bjorn was suggesting...at least in my opinion. 

I added a TODO (with a jira ticket) to just investigate if it makes sense / is possible to not mutably expose the hashmap at a later time. In the past, I've found that this pattern can lead to problems but I understand there may be no reasonable way around it.

Reply:
To be honest, if it were me, I would just disable the type complexity clippy error for the whole repository.
Sometimes types are simply complex and using a `type` is basically just a workaround and requiring you to search for the `type` definition to just see the actual type.